File: app/src/main/java/com/example/msimple/data/model/CategoryModel.java
package com.example.msimple.data.model;

import com.example.msimple.Delegate.CategoryDelegate;
import com.example.msimple.data.vos.CategoryVO;
import com.example.msimple.data.vos.ProgramVO;
import com.example.msimple.utils.AppConstants;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class CategoryModel extends BaseModel implements ICategoryModel {

    Map<String, ProgramVO>mData;
    private static CategoryModel mCategoryModel;

    private CategoryModel() {
        mData = new HashMap<>();
    }

    public static CategoryModel getInstance() {
        if (mCategoryModel == null) {
            mCategoryModel = new CategoryModel();
        }
        return mCategoryModel;
    }

    @Override
    public void getCategory(final Response response) {

        mRetrofit.loadCategory(AppConstants.ACCESS_TOKEN, "1", new CategoryDelegate() {
            @Override
            public void onSuccess(List<CategoryVO> categoryVO) {
                response.onSuccess(categoryVO);
                for(int i=0;i<categoryVO.size();i++)
                {
                    for (int j=0;j<categoryVO.get(i).getPrograms().size();j++)
                    {
                        mData.put(categoryVO.get(i).getPrograms().get(j).getProgramId(),categoryVO.get(i).getPrograms().get(j));
                    }
                }
            }

            @Override
            public void fail(String message) {

                response.onFail(message);
            }
        });

    }

    @Override
    public ProgramVO getProgramVO(String id) {

        return mData.get(id);
    }
}

File: app/src/main/java/com/example/msimple/data/model/CategoryModel.kt
package com.example.msimple.data.model

import com.example.msimple.delegate.CategoryDelegate
import com.example.msimple.data.vos.CategoryVO
import com.example.msimple.data.vos.ProgramVO
import com.example.msimple.utils.AppConstants

import java.util.HashMap

class CategoryModel private constructor() : BaseModel(), ICategoryModel {

    internal var mData: MutableMap<String, ProgramVO>

    init {
        mData = HashMap()
    }

    override fun getCategory(response: ICategoryModel.Response) {

        mRetrofit.loadCategory(AppConstants.ACCESS_TOKEN, "1", object : CategoryDelegate {
            override fun onSuccess(categoryVO: List<CategoryVO>) {
                response.onSuccess(categoryVO)
                for (i in categoryVO.indices) {
                    for (j in 0 until categoryVO[i].programs.size) {
                        mData[categoryVO[i].programs[j].programId] = categoryVO[i].programs[j]
                    }
                }
            }

            override fun fail(message: String) {

                response.onFail(message)
            }
        })

    }

    override fun getProgramVO(id: String): ProgramVO? {

        return mData[id]
    }

    companion object {
        private var mCategoryModel: CategoryModel? = null

        val instance: CategoryModel
            get() {
                if (mCategoryModel == null) {
                    mCategoryModel = CategoryModel()
                }
                return mCategoryModel
            }
    }
}
